### PR TITLE
Tool selection capabilities

### PR DESCRIPTION
**Pull request by raulraja**
```java
package com.xebia.functional.xef.java.auto.reasoning;

import com.xebia.functional.xef.auto.CoreAIScope;
import com.xebia.functional.xef.auto.llm.openai.OpenAI;
import com.xebia.functional.xef.auto.llm.openai.OpenAIEmbeddings;
import com.xebia.functional.xef.reasoning.filesystem.Files;
import com.xebia.functional.xef.reasoning.pdf.PDF;
import com.xebia.functional.xef.reasoning.text.Text;
import com.xebia.functional.xef.reasoning.tools.ToolSelection;

import java.util.Collections;
import java.util.List;

public class ToolSelectionExample {

    public static void main(String[] args) {
        var scope = new CoreAIScope(new OpenAIEmbeddings(OpenAI.DEFAULT_EMBEDDING));
        var model = OpenAI.DEFAULT_CHAT;
        var serialization = OpenAI.DEFAULT_SERIALIZATION;
        var text = Text.create(model, scope);
        var files = Files.create(serialization, scope, Collections.emptyList());
        var pdf = PDF.create(model, serialization, scope);

        var toolSelection = new ToolSelection(
                serialization,
                scope,
                List.of(
                    text.summarize,
                    pdf.readPDFFromUrl,
                    files.readFile,
                    files.writeToTextFile
                ),
                Collections.emptyList()
        );

        var inputText = "Extract information from https://arxiv.org/pdf/2305.10601.pdf";
        var result = toolSelection.applyInferredToolsBlocking(inputText);
        System.out.println(result);
    }
}
```
```txt
I 22:46:15.461 [.reasoning.tools.ToolSelection] 🔍 Applying inferred tools for task: Extract information from https://arxiv.org/pdf/2305.10601.pdf
I 22:46:15.461 [.reasoning.tools.ToolSelection] 🔍 Creating execution plan for task: Extract information from https://arxiv.org/pdf/2305.10601.pdf
I 22:46:24.569 [.reasoning.tools.ToolSelection] 🔍 Applying execution plan with reasoning: This execution plan ensures that the PDF is read, the information is summarized, and the summary is written to a text file
I 22:46:24.569 [.reasoning.tools.ToolSelection] 🔍 Applying tool: Read PDF from URL for step: The task requires extracting information from a PDF file located at a specific URL
I 22:46:26.346 [f.reasoning.pdf.ReadPDFFromUrl] Reading url https://arxiv.org/pdf/2305.10601.pdf
I 22:46:28.475 [oning.text.summarize.Summarize] 📚 Summarizing large text of tokens 55971 to approximately Bound(length=1000) tokens
I 22:46:28.678 [oning.text.summarize.Summarize] 📚 Split text into 5 chunks
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 5145 for length Bound(length=200)
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 14385 for length Bound(length=200)
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 7644 for length Bound(length=200)
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 12421 for length Bound(length=200)
I 22:46:28.682 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 14388 for length Bound(length=200)
I 22:46:29.992 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 34
I 22:46:29.992 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 26
I 22:46:30.034 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 18
I 22:46:31.229 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 114
I 22:46:32.756 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 107
I 22:46:32.757 [oning.text.summarize.Summarize] 📚 Summarized 5 chunks
I 22:46:32.757 [.reasoning.tools.ToolSelection] 🔍 Applying tool: Summarize for step: After reading the PDF, we need to summarize the extracted information
I 22:46:32.757 [oning.text.summarize.Summarize] 📚 Summarizing large text of tokens 324 to approximately Bound(length=1000) tokens
I 22:46:32.758 [oning.text.summarize.Summarize] 📚 Split text into 1 chunks
I 22:46:32.758 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 327 for length Bound(length=1000)
I 22:46:35.507 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 101
I 22:46:35.507 [oning.text.summarize.Summarize] 📚 Summarized 1 chunks
I 22:46:35.507 [.reasoning.tools.ToolSelection] 🔍 Applying tool: Write output to Text file for step: Finally, we need to write the summarized information to a text file

├── Read PDF from URL
│   ├── The task requires extracting information from a PDF file located at a specific URL
│   └── The "Tree of Thoughts" (ToT) framework is introduced to enhance language models' problem-solving abilities. It allows language models to consider multiple reasoning paths and make deliberate decisions by exploring coherent units of text ("thoughts"). ToT significantly improves performance on tasks requiring planning or search. Existing approaches like Chain-of-Thought (CoT) prompting are limited in their exploration and decision-making capabilities. ToT addresses these limitations by decomposing the problem-solving process into thought steps, generating potential thoughts, evaluating states, and using search algorithms. The given text is too long to be summarized within a maximum bound of 200 words. I'm sorry, but the provided text does not seem to be readable or meaningful. It appears to be a string of random characters. Therefore, it cannot be summarized. 1 - The text provided is not readable and does not contain any meaningful information. Therefore, it cannot be summarized.

2 -  1 - The Tree of Thoughts (ToT) framework enhances the problem-solving capabilities of language models (LMs) by empowering them to make autonomous and intelligent decisions. However, future applications involving interaction with external environments or humans could pose potential dangers. On the other hand, ToT improves the interpretability of model decisions and the opportunity for human alignment. LMs address the weakness of classical methods and provide a way to solve complex problems that are not easily formalized. The intersection of LMs with classical approaches to AI is an exciting direction for future work.

2 - 

├── Summarize
│   ├── After reading the PDF, we need to summarize the extracted information
│   └── 1 - The "Tree of Thoughts" (ToT) framework enhances language models' problem-solving abilities by allowing them to consider multiple reasoning paths and make deliberate decisions. ToT improves performance on tasks requiring planning or search by decomposing the problem-solving process into thought steps, generating potential thoughts, evaluating states, and using search algorithms. Existing approaches like Chain-of-Thought (CoT) prompting have limitations in exploration and decision-making capabilities. ToT addresses these limitations and provides a more comprehensive problem-solving approach.

├── Write output to Text file
│   ├── Finally, we need to write the summarized information to a text file
│   └── /var/folders/h7/vh204ss54dv8ghllkvrkhlxr0000gn/T/3e7c7b98-840d-405e-9236-251b4f553b17/summary.txt

### CODE REVIEW

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/tools/ToolSelection.kt [hunk 1]
```diff
@@ -3,45 +3,112 @@ package com.xebia.functional.xef.reasoning.tools
...
+
+  override suspend fun invoke(input: String): String {
+    return when (val trace = applyInferredTools(input)) {
+      is ToolsExecutionTrace.Completed -> trace.output
+      ToolsExecutionTrace.Empty -> ""
+    }
+  }
+
+  @JvmBlocking
+  @JvmAsync
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/tools/ToolSelection.kt:31]: @nomisRev This seems to do similar to what we were attempting to build with `@AsFuture` .
This creates an async and sync method and makes the suspend method @JvmSynthetic. https://github.com/ForteScarlet/kotlin-suspend-transform-compiler-plugin

**Comment by raulraja** [comment]
@juanpedromoreno this seem to work for me in my local build but fails here with https://github.com/xebia-functional/xef/actions/runs/5638796985/job/15273897939 

```
> Task :xef-core:compileKotlinJvm
e: file:///home/runner/work/xef/xef/core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt:15:11 Unresolved reference: github
e: file:///home/runner/work/xef/xef/core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt:115:18 Unresolved reference: KotlinLogging
```
```
* What went wrong:
Execution failed for task ':buildAndTestSinglep'.
```
This setup seems not to work when multiple multiplatform dependencies are in play. In this case, we depend on xef fileSystem for JVM in the reasoning module, but not sure if it's related.

File: reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/pdf/ReadPDFFromFile.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.xebia.functional.xef.reasoning.text.summarize.Summarize
+import com.xebia.functional.xef.reasoning.text.summarize.SummaryLength
+import com.xebia.functional.xef.reasoning.tools.Tool
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.io.File
+import kotlinx.serialization.Serializable
+
+@Serializable data class ExtractedPDFFile(val absolutePath: String)
+
+class ReadPDFFromFile(
```
Comment (nomisRev) [inline reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/pdf/ReadPDFFromFile.kt:17]:
```suggestion
class ReadPDFFromFile @JvmOverloads constructor(
```

File: reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/pdf/ReadPDFFromUrl.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import com.xebia.functional.xef.prompt.Prompt
+import com.xebia.functional.xef.reasoning.text.summarize.Summarize
+import com.xebia.functional.xef.reasoning.text.summarize.SummaryLength
+import com.xebia.functional.xef.reasoning.tools.Tool
+import io.github.oshai.kotlinlogging.KotlinLogging
+import kotlinx.serialization.Serializable
+
+@Serializable data class ExtractedUrl(val url: String)
+
+class ReadPDFFromUrl(
```
Comment (nomisRev) [inline reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/pdf/ReadPDFFromUrl.kt:16]:
```suggestion
class ReadPDFFromUrl @JvmOverloads constructor(
```

File: reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/search/Search.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import com.xebia.functional.xef.agents.search
+import com.xebia.functional.xef.auto.CoreAIScope
+import com.xebia.functional.xef.llm.Chat
+import com.xebia.functional.xef.reasoning.text.Text
+import com.xebia.functional.xef.reasoning.text.summarize.Summarize
+import com.xebia.functional.xef.reasoning.text.summarize.SummaryLength
+import com.xebia.functional.xef.reasoning.tools.LLMTool
+import com.xebia.functional.xef.reasoning.tools.Tool
+
+class Search(
```
Comment (nomisRev) [inline reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/search/Search.kt:12]:
```suggestion
class Search @JvmOverloads constructor(
```

**Pull request merged**